vowpalwabbit/parser.cc

```diff
@@ -132,33 +126,28 @@ void reset_source(vw& all, size_t numbits)
 {
   io_buf* input = all.p->input;
   input->current = 0;
+
+  // If in write cache mode then close all of the input files then open the written cache as the new input.
   if (all.p->write_cache)
   {
     all.p->output->flush();
+    // Turn off write_cache as we are now reading it instead of writing!
     all.p->write_cache = false;
     all.p->output->close_file();
+
+    // This deletes the file from disk.
     remove(all.p->output->finalname.begin());
 
+    // Rename the cache file to the final name.
     if (0 != rename(all.p->output->currentname.begin(), all.p->output->finalname.begin()))
       THROW("WARN: reset_source(vw& all, size_t numbits) cannot rename: " << all.p->output->currentname << " to "
                                                                           << all.p->output->finalname);
-
-    while (input->num_files() > 0)
-      if (input->compressed())
-        input->close_file();
-      else
-      {
-        int fd = input->files.pop();
-        const auto& fps = all.final_prediction_sink;
-
-        // If the current popped file is not in the list of final predictions sinks, close it.
-        if (std::find(fps.cbegin(), fps.cend(), fd) == fps.cend())
-          io_buf::close_file_or_socket(fd);
-      }
-    input->open_file(all.p->output->finalname.begin(), all.stdin_off, io_buf::READ);  // pushing is merged into
-                                                                                      // open_file
+    input->close_files();
+    // Now open the written cache as the new input file.
+    input->add_file(VW::io::open_file_reader(all.p->output->finalname.cbegin()));
```

Comment:
Could is_resettable/reset() here instead of open.

Reply:
This check is in the block for `if (all.p->write_cache)`. That means a cache file has just been written (from text), so it is closing the old text file and reopening the just written cache file I believe.